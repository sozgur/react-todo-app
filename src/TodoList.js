import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";
import { v4 as uuidv4 } from "uuid";
import { useState } from "react";

const TodoList = () => {
    const INITIAL_STATE = [];
    const [todos, setTodos] = useState(INITIAL_STATE);

    const addTodo = (newTodo) => {
        setTodos([...todos, { ...newTodo, id: uuidv4() }]);
    };

    const removeTodo = (id) => {
        setTodos(todos.filter((todo) => id !== todo.id));
    };

    return (
        <div>
            <NewTodoForm addTodo={addTodo} />
            {todos.map(({ id, title }) => (
                <Todo key={id} id={id} title={title} remove={removeTodo} />
            ))}
        </div>
    );
};

export default TodoList;
