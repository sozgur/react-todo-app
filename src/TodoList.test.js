import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

function addTodo(todoList, title = "clean car!") {
    const titleInput = todoList.getByLabelText("Title");
    fireEvent.change(titleInput, { target: { value: title } });
    const submitTodo = todoList.getByText("Add");
    fireEvent.click(submitTodo);
}

it("renders without crashing", () => {
    render(<TodoList />);
});

it("matches snapshot", () => {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});

it("can add a todo", () => {
    const todoList = render(<TodoList />);
    addTodo(todoList);

    expect(todoList.getByText("X")).toBeInTheDocument();
    expect(todoList.getByLabelText("Title")).toHaveValue("");
});

it("can delete a todo", () => {
    const todoList = render(<TodoList />);
    addTodo(todoList);

    const deleteButton = todoList.getByText("X");
    fireEvent.click(deleteButton);

    expect(todoList.queryByText("clean car!")).not.toBeInTheDocument();
});

it("can complete a todo", () => {
    const todoList = render(<TodoList />);
    addTodo(todoList);

    const todo = todoList.queryByText("clean car!");

    expect(todoList.queryByText("clean car!")).not.toHaveClass("checked");

    fireEvent.click(todo);

    expect(todoList.queryByText("clean car!")).toHaveClass("checked");
});
