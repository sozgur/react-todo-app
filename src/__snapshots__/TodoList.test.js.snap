// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <h2>
        My To Do List
      </h2>
      <form>
        <label
          for="title"
        >
          Title
        </label>
        <input
          id="title"
          name="title"
          type="text"
          value=""
        />
        <button>
          Add
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
